libsel4vmmplatsupport/src/arch/arm/psci.c

```diff
@@ -50,23 +50,37 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
         smc_set_return_value(&regs, 0x00010000); /* version 1 */
         break;
     case PSCI_CPU_ON: {
-        uintptr_t target_cpu = smc_get_arg(&regs, 1);
+        uintptr_t requested_cpu = smc_get_arg(&regs, 1);
         uintptr_t entry_point_address = smc_get_arg(&regs, 2);
         uintptr_t context_id = smc_get_arg(&regs, 3);
-        vm_vcpu_t *target_vcpu = vm_vcpu_for_target_cpu(vcpu->vm, target_cpu);
-        if (target_vcpu == NULL) {
-            target_vcpu = vm_find_free_unassigned_vcpu(vcpu->vm);
-            if (target_vcpu && start_new_vcpu(target_vcpu, entry_point_address, context_id, target_cpu) == 0) {
+        vm_vcpu_t *target_vcpu = NULL;
+        if ((requested_cpu >= 0) && (requested_cpu < vcpu->vm->num_vcpus)) {
+            target_vcpu = vcpu->vm->vcpus[requested_cpu];
+        } else {
+            smc_set_return_value(&regs, PSCI_INTERNAL_FAILURE);
+            break;
+        }
+
+        /* Automatically assign vcpu to an unassigned physical cpu */
+        if (target_vcpu->target_cpu == -1) {
+            int selected_cpu = vm_find_free_unassigned_cpu(vcpu->vm);
+            if ((selected_cpu >= 0) && start_new_vcpu(target_vcpu, entry_point_address, context_id, selected_cpu) == 0) {
                 smc_set_return_value(&regs, PSCI_SUCCESS);
             } else {
                 smc_set_return_value(&regs, PSCI_INTERNAL_FAILURE);
             }
-        } else {
+        } else if ((target_vcpu->target_cpu >= 0) && (target_vcpu->target_cpu < CONFIG_MAX_NUM_NODES)) {
+            /* Assign vcpu to physical cpu specified in config */
             if (is_vcpu_online(target_vcpu)) {
                 smc_set_return_value(&regs, PSCI_ALREADY_ON);
+            } else if (start_new_vcpu(target_vcpu, entry_point_address, context_id, target_vcpu->target_cpu) == 0) {
+                smc_set_return_value(&regs, PSCI_SUCCESS);
             } else {
                 smc_set_return_value(&regs, PSCI_INTERNAL_FAILURE);
             }
+        } else {
+            ZF_LOGE("Target CPU was an invalid number");
```

Comment:
We should also log the invalid number
```suggestion
            ZF_LOGE("Target CPU is invalid: %d", target_vcpu->target_cpu);
```

Reply:
Done